### ISSUE TITLE
Update R package to v3.1.0 (beta3)

### ISSUE THREAD

**Issue by steffilazerte** [MotusDev/Motus-TO-DO#717]
@denislepage, @cjardine-bsc, @StuMackenzie, and @jsayers-bsc (I'm sure there are others, but I'm not sure whom I should spam)

I think we're about ready to update the motus package with all the small tweaks that have been made on the beta3 branch. This would update the package from v3.0.0 to v3.1.0.

Namely this would add:

**Small Changes**
* Receiver downloads now have similar progress messages to Project downloads
* Moved GPS fields from `alltags` to `alltagsGPS`
* New function `getGPS()` adds GPS fields to data
* Remove NOT NULL constraint on `motusTagID`s in ambiguous tag view

**Bug fixes**
* Receivers updating to the new version errored on the download start
* Allow renaming (if possible) of large databases on data updates
* Continue checking for activity/nodeData even if first batch returns 0
* `nodeDataId` is corrected to `nodeDataID`

I don't believe any of these changes rely on changes to the API, and all my tests pass on both the beta and normal API.

### To proceed:
- [x] Do people agree that it's time to update?
- [x] Can I have a couple of volunteers to install motus from the beta branch and just run a couple of things to double check?
    - [x] Is the new `alltags` field better and faster than the old `alltags` fields?
    - [x] `getGPS()` seem to do what it is supposed to?
    - [x] No new errors?
- [x] I will make updates to the Motus R Book that correspond to these new changes
- [ ] I'll merge beta3 to master!

**Comment by steffilazerte** [comment]
I've updated the [beta branch](https://github.com/MotusWTS/MotusRBook/tree/beta) of the Motus R book to reflect the changes in motus v3.1.0

**Comment by steffilazerte** [comment]
Hi @jsayers-bsc thanks for looking into this for me!

1. NOT NULL CONSTRAINT is definitely related to a fix I did with the nodeDataID... I had trouble fixing it (I only half fixed it the first time, but hopefully it's completely fixed now). Hopefully that was a blip that is now not an issue but let me know if it pops up (it would only happen to projects that have `nodeData`, I think).

2. Glad to see the performance actually increased!

3. Yes, getGPS returns a flat data frame, which is why it needs to be assigned. I'll make that clearer in the documentation. This is supposed to help you add in GPS data on a subset, because if you wanted the entire data frame, you'd just use `alltagsGPS`, right?

That's a good point about omitting records that don't have a GPS hit. It should be fairly straightforward to add an option to keep all records. I'll look in to that!

**Comment by leberrigan** [comment]
I just updated my package to the beta3 and  had the following issues:
 - Database that already existed (last updated in March) gave me this error:
```
updateMotusDb started (3 version update(s))
 - Move GPS data from alltags view to alltagsGPS view
 - Drop IS NOT NULL constraint on allambigs
Error in FUN(X[[i]], ...) : 
  Error : error in view alltags_fast: no such table: main.allambigs
```
- Deleting/renaming the old database fixed the issue, but now I have to download the data again. Is there not a way to just update the existing database?

**Comment by steffilazerte** [comment]
It was supposed to work without requiring a re-download. I'll take a look. Do you still have the original data base that you could test on?

**Comment by steffilazerte** [comment]
@leberrigan, I think this may be related to the fact that you added a custom view (alltags_fast) that depended on another view ("allambigs"). When I updated the SQLite file to Drop the IS NOT NULL constraint on "allambigs", I needed to update the view (i.e. remove and recreate it with edits). 

I think if you start with your original database, delete your custom view ("alltags_fast"), and then open it in with `tagme()` it should update without problems. **Does it?**

How common are custom views that depend on motus views? If they're very common, perhaps I should add a message to the update script warning users if motus views are going to be altered (and thereby potentially disruption custom views).

**Comment by leberrigan** [comment]
Aha, that's it! It works without issue if I remove that view first using the following code:

```
sql <- DBI::dbConnect(RSQLite::SQLite(), paste0(dir, "project-", projID, ".motus"))
dropview <- "DROP VIEW alltags_fast"
DBI::dbSendQuery(sql, dropview)
DBI::dbDisconnect(sql.test)
```

It's going to be fairly common for folks to have this since we've had to post this as the standard fix after receiving a bunch of complaints that the databases took much to long to load with the last update.

**Comment by steffilazerte** [comment]
Okay that's good to know... I'll think about how to deal with this.

Perhaps a warning, if non-standard views are detected, espcially in this case, if they generally have the name "alltags_fast"? If they're all likely to be "alltags_fast", I could preemptively remove that view, but I don't think it's polite to mess around with users customization.

I'll see if I can't get something figured out soon.

**Comment by denislepage** [comment]
I suspect the main issue will be that particular name. I circulated some custom code earlier that was based on that name. Are you able to make it ignore the step of creating the new view if one already exists, or create it with a different name and warn the user? E.g. append a 2 as in alltags_fast2. If they already have that as well, you could go to alltag_fast3, etc.

That would be the ideal approach. Deleting a custom view without the user consent is not ideal, agreed.

**Comment by steffilazerte** [comment]
I can definitely rename the table if needed, but the problem is that if the view is based on a table I need to delete in order to update, there'll be an error (I had to change a couple of IF NULL CONSTRAINTS, which means I had to re-create tables).

I'm seeing if I can figure out which tables a view depends on. I could then check custom views to see if they depend on something that will be deleted/recreated and either error and stop, or ask the user if they want to delete the view...

**Comment by steffilazerte** [comment]
Otherwise, I can let the error happen, and try to catch it gracefully to let the user know what's going on

**Comment by denislepage** [comment]
Ah, yes. Perhaps you can see if you can read the view definition, store it and recreate it afterwards? That’s a bit risky that people will lose their view if something fails.

**Comment by steffilazerte** [comment]
As I see it there are about three solutions. In all cases, motus checks if there are any custom views that depend on a motus views that will be temporarily removed (no point in making people remove views if there won't be a problem). Then motus can do one of the following:

1. Stop with an error (currently implemented; @leberrigan you should now get an informative error)
```
 Error: This SQLite database contains some custom views (alltags_fast) which will have
       to be removed before the update can proceed.
```

2. Ask for intervention:
```
This SQLite database contains some custom views (alltags_fast) which will have
to be removed before the update can proceed. Allow motus to delete them? 
(The SQLite commands have been saved to custom_views_2020-05-14.log)
1. Yes (continue with update)
2. No (stop)
```

3. Re-implement the views:
```
This SQLite database contains some custom views (alltags_fast) which will have
to be removed before the update can proceed. motus will remove these views and 
then try to recreate them. Allow motus to proceed?
1. Yes (continue with update)
2. No (stop)
```

My preference is for the first or second option. I don't like the third because I think it's too fragile. I would however, like to add some instructions for how to remove the view. Generally I'd assume that people who can create a view can delete it. But I don't think that's a safe assumption in this case.

I'm not really sure where some of these trouble shooting tips should go. They could go in the Motus R book under "Troubleshooting", and then the error message could contain a link. But we'll have to make sure the Book and the package are in sync.

```
 Error: This SQLite database contains some custom views (alltags_fast) which will have
       to be removed before the update can proceed. 
       See https://motus.org/MotusRBook/appendixB.html#customviews for examples
```

Alternatively, simple fixes like this one in the package documentation under "help". But it may get confusing to have help in multiple places.

```
 Error: This SQLite database contains some custom views (alltags_fast) which will have
       to be removed before the update can proceed. 
       Type ?help and look under the "Deleting custom views" section for examples.
```

**Comment by denislepage** [comment]
I think I would go with option 2 if that’s straightforward. Otherwise option 1. Perhaps add a note saying that if they want to keep their views, they should say No and create the custom views under another name.

**Comment by steffilazerte** [comment]
Okay, the update function now asks the users if it can delete the view before proceeding. No means stop, yes means delete and update (sql code is saved to log file so the user can recreate the view if they want).

The problem isn't actually the name of the view, it's if a custom view depends on a motus View that I have to delete (in order to make a change). Then it gets annoyed. 

This is on beta3 and ready to go.

**Comment by steffilazerte** [comment]
@jsayers-bsc I clarified the examples in `getGPS()` to emphasize that you need to save the output to a an object and I add an argument `keepAll`. When TRUE, this will return all hits regardless of whether there was a GPS match or not. When FALSE, it returns only those with GPS matches. (This is on beta3 now)

**Comment by denislepage** [comment]
Are we ready to deploy to live then? Steffi, are you able to post a summary of the changes like you did in a previous instance?

**Comment by steffilazerte** [comment]
Sorry for the delay @denislepage. It's been a stupid couple of days.

I finished the last couple of tweaks (mostly on the tests, nothing on the actually motus functions). 

I think we're ready unless any one found more problems, or problems with my fixes.  The NEWS.md file has a summary of the change (copied below).

Give me the go ahead and I'll merge with master, and then tag the release with this information. I assume you'll want to send out an email to that effect too?

------------- 

# motus 3.1.0

### Small Changes
* Receiver downloads now have similar progress messages to Project downloads
* Move GPS fields from `alltags` to `alltagsGPS`
* New function `getGPS()` adds GPS fields to data
* Remove NOT NULL constraint on `motusTagID`s in ambiguous tag view

### Bug fixes
* Receivers updating to the new version errored on the download start
* Allow renaming (if possible) of large databases on data updates
* Continue checking for activity/nodeData even if first batch returns 0
* `nodeDataId` is corrected to `nodeDataID`
* Warn users with custom views if they need to be removed prior to updating

**Comment by denislepage** [comment]
Thanks.

I think a few of these new things should be documented in the R book as well. Particularly the new GPS changes affecting the view and the getGPS function. The note about getGPS function is a bit cryptic in this, and we should probably refer users to the R book for more details.

I’ve started to make some edits to the chapter 3, which should be looked at and tested. I’ll let you all know when I have posted my changes.

I’m also wondering if we can rename the alltags parameter in the getGPS function? People may want to use this with a custom subset of data that is far removed from the alltags view. This parameter can be either a dataframe or a sqlite object, correct?

**Comment by denislepage** [comment]
Sorry, I was working on the wrong branch of motusRBook, and just noticed that you had already made some changes. I am not yet seeing anything about getGPS in chapter 3 however, so I'll copy over my changes to the beta branch. When merging into master later, you can ignore the changes already in master if git complains.

**Comment by steffilazerte** [comment]
In chapter 3, the users are introduced to the views `alltags` and `alltagsGPS`. The `getGPS()` function is discussed in Chapter 5 (there is a [reference in Chapter 3 to this](https://beta.motus.org/MotusRBook/accessingData.html#accessSQL)) as I thought it was more of a special case for when users are filtering their data. If a user isn't going to first filter their data, there's really no reason to use the `getGPS()` function instead of the `alltagsGPS` view.

And yes, I can change the name of `alltags` to something else. It either has to be a data frame with "hitID", "batchID" and "ts" OR the pointer to a SQLite src. I'm not sure there's 100% consistency, but other functions that take a similar object use the argument name `data`. Should I implement that here?

**Comment by denislepage** [comment]
I tested the new branch. Seems all to be working fine. One potential issue of the getGPS function is tha the user has no control over the time interval  between the hit and the GPS fix. When I tried it, it gave me:

`Max time difference between GPS location and hit is: 244936.11 min`

That is 170 days (I stopped the tagme function, so I don't know if my data was complete).

The most flexible option would be to add a gpsTs value in the return dataframe, for all 3 methods. An optional feature could be to include a max interval parameter, which would prevent distant (in time) GPS fixes from being used with the "closest" option (e.g. maxInterval = 1440 minutes).

Adding the gps timestamp is probably a good first step if that can be done in beta3. The additional parameter could also be included, if that's easy, or we can wait until the next release, as users can do that part themselves using the comparison between ts in hits in gpsTs.

**Comment by denislepage** [comment]
I pushed some changes to chapter 3 in beta (and partly reverted my changes in master branch), but you can see how they clash with your edits. Feel free to discard things, but I would at least leave a reference to the GPS function in the same location after people load the alltags view, or we can leave it in both places. Loading GPS data is perhaps more relevant to chapter 3 than 5, so we can at least cover the option of loading GPS data there.

On the max ts interval, we could probably leave the default as null. The correct value to use might depend on each receiver, and there may not be a one-size-fits-all. People with more customized needs can do their own using gpsTs.

**Comment by denislepage** [comment]
Yes, I was thinking we can use data there too.

**Comment by steffilazerte** [comment]
Hi @denislepage I'm just working in the changes to `getGPS()`.

Looking at my code, I think I may have been restricting the matching too much. Originally I had it so that a GPS point could only match hitIDs within the same batch. But I think that restriction is no good. They should be able to match hitIDs no matter the batch, correct? Especially if we're taking the median location within 24hrs, that could definitely span several batches. Or if we take the closest GPS point in time, for a hit at the start of one batch, that could be a GPS point at the end of the previous batch...

So I think I should remove the within-batchID restriction. I wanted to check in and make sure this seems sensible to you before I make too many changes.

**Comment by denislepage** [comment]
Yes, we discussed this today. Looking within batch is probably too restrictive, especially for networked receivers that have short batches (e.g. 1 hour long)

**Comment by steffilazerte** [comment]
Something else has just popped up. Lately the dbplyr and dplyr package don't require a special connection object (i.e., `dbplyr::src_dbi()`) to connect to the database. They can use the one created by the DBI package (i.e. `DBI::dbConnect()`).

We don't have to fix this (dplyr deprecated a function, but the one we use is in the dbplyr. It is listed as superseded, not deprecate). But I think we should eventually for two reasons:

1. It may eventually be deprecated by dbplyr
2. It would be nice to have consistent behaviour between database connections returned by the `tagme()` function and `DBI::dbConnect()`. Right now, if you want to use the DBI package on a tag me function, you need to specify the connection with `mydata$con`, but with DBI you only use `mydata`. By updating motus, we would be consistently using `mydata` everywhere. 

This is actually a relatively straightforward fix. Instead of creating the connection like:

```
rv <- dbplyr::src_dbi(con = DBI::dbConnect(RSQLite::SQLite(), dbname))
```

We would remove the deprecated function and just retain the direct connection:

```
rv <- DBI::dbConnect(RSQLite::SQLite(), dbname)
```

However, I will have to update all the functions to use `rv` rather than `rv$con` accordingly. 

Further, users will have to change their code from using `rv$con` to just `rv`. This means that how users interact with the database will change (for the better, but still).

I would suggest that when we make this change, I insert a message into the `tagme()` function that alerts users to this change.

So what do you think? Is this worth implementing now or should we wait until the next iteration?

**Comment by steffilazerte** [comment]
@denislepage @StuMackenzie @jsayers-bsc @cjardine-bsc 

I want to see if we can't wrap things up here!

1. Denis' response via email was to put the above issue (regarding database connections) on hold for now. 

2. Regarding [Denis' comments (above)](https://github.com/MotusDev/Motus-TO-DO/issues/717#issuecomment-635441108) for the Motus R book:

> I pushed some changes to chapter 3 in beta (and partly reverted my changes in master branch), but you can see how they clash with your edits. Feel free to discard things, but I would at least leave a reference to the GPS function in the same location after people load the alltags view, or we can leave it in both places. Loading GPS data is perhaps more relevant to chapter 3 than 5, so we can at least cover the option of loading GPS data there.

I edited Chapter 3 so there is now solid information on using the `getGPS()` function in both Chapters 3 and 5 (Chapter 5 also refers the users to the slightly more comprehensive examples in Chapter3).

---

As a result, I think we're ready to update. I've re-run the tests and everything still checks out. The only thing is [what I mentioned several comments above](https://github.com/MotusDev/Motus-TO-DO/issues/717#issuecomment-638973246): people may want to double check the changes to the `getGPS()` function and the beta branch of the motus R book first. 

**Can I update?**

**Comment by denislepage** [comment]
Fine for me. We can fix issues if they arise. I am ready to FTP the R book on the web site? Has the beta branch been merged in master?

**Comment by steffilazerte** [comment]
Neither of the beta branches have been merged into master (motus and motusRbook). I wanted to get the update go ahead first. I'll merge with master then get back to you!

**Comment by steffilazerte** [comment]
I've created pull requests and will merge them in as soon as the CI tests finish (in an abundance of caution!). Then I'll update the sandbox branch to reflect the new changes and get a clean beta3 branch (in case we have more to release as motus v3.2.0 before we get to v4.0.0)

**Comment by steffilazerte** [comment]
Okay, I've merged beta3 with master so the master branch is now v3.1.0 (I've also logged a release here: https://github.com/MotusWTS/motus/releases/tag/v3.1.0)

Motus R book is also updated to use the newest version and have the additional details.

In both cases I deleted the beta branches and create new ones from master. I still have to update sandbox from master but I'll get to that later. 

Go serve the Motus R book!

News are here: https://github.com/MotusWTS/motus/blob/master/NEWS.md and logged in the release as well.
